dali/pipeline/operators/arithmetic/expression_tree.cc

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "dali/core/format.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_cpu.h"
+#include "dali/pipeline/operators/arithmetic/expression_tree.h"
+
+namespace dali {
+
+namespace {
+
+using ParseResult = std::tuple<std::unique_ptr<ExprNode>, int>;
```

Comment:
What's the benefit having `tuple` over `struct`? I think, that it's good rule of thumb, that using `tuple` for known types is pointless

Reply:
std::tie.